c_reference/src/quantized_fastgrnn.c

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+#include "quantized_fastgrnn.h"
+
+int q_fastgrnn_lr(MYINT* const hiddenState, MYITE hiddenDims,
+                  const MYINT* const input, MYITE inputDims, MYITE steps,
+                  const void* params, void* buffers, const void *scales,
+                  int backward, int normalize) {
+  const Q_FastGRNN_LR_Params* tparams = (const Q_FastGRNN_LR_Params*)params;
+  Q_FastGRNN_LR_Buffers* tbuffers = (Q_FastGRNN_LR_Buffers*)buffers;
+  const Q_FastGRNN_LR_Scales* tscales = (const Q_FastGRNN_LR_Scales*)scales;
+
+  if (tbuffers->preComp1 == 0) return ERR_PRECOMP_NOT_INIT;
+  if (tbuffers->preComp2 == 0) return ERR_PRECOMP_NOT_INIT;
+  if (tbuffers->preComp3 == 0) return ERR_PRECOMP_NOT_INIT;
+  if (tbuffers->tempLRW == 0) return ERR_TEMPLRW_NOT_INIT;
+  if (tbuffers->tempLRU == 0) return ERR_TEMPLRU_NOT_INIT;
+  if (tbuffers->normFeatures == 0) return ERR_NORMFEATURES_NOT_INIT;
+
+  // #steps iterations of the RNN cell starting from hiddenState
+  for (MYITE t = 0; t < steps; t++) {
+    // Normalize the features
+    MYITE offset = backward ? steps - 1 - t : t;
+    if (normalize) {
+      // This diverges from the original implementation because of
+      // impracticality of scaled addition beyond 0, 1, and -1 multipliers
+      v_q_sub(input + offset * inputDims, tparams->mean + t * inputDims,
+              inputDims, tbuffers->normFeatures, tscales->input,
+              tscales->mean, tscales->mean_sub);
```

Comment:
This piece of code follows from the master branch code. I believe `t` should be replaced by offset, both here and in the original code, unless I'm misunderstanding something.